internal/crdmetrics/tempo_stack_test.go

```diff
@@ -0,0 +1,182 @@
+		if expectedAttrSet.Equals(&dp.Attributes) {
+			matchingDP = dp
+			found = true
+			break
+		}
+	}
+
+	assert.True(t, found, "Metric %s doesn't have expected attributes %v", expectedAttrs)
+	assert.Equal(t, expectedValue, matchingDP.Value,
+		"Metric %s doesn't have expected value %d, got %d", name, expectedValue, matchingDP.Value)
+}
+
+func newTempoStackInstance(nsn types.NamespacedName, managedState v1alpha1.ManagementStateType,
+	storage v1alpha1.ObjectStorageSecretType) v1alpha1.TempoStack {
+	return v1alpha1.TempoStack{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nsn.Name,
+			Namespace: nsn.Namespace,
+		},
+		Spec: v1alpha1.TempoStackSpec{
+			ManagementState: managedState,
+			Storage: v1alpha1.ObjectStorageSpec{
+				Secret: v1alpha1.ObjectStorageSecretSpec{
+					Type: storage,
+				},
+			},
+		},
+	}
+}
+
+func newExpectedMetric(name string, keyPair attribute.KeyValue, value int64) expectedMetric {
+	return expectedMetric{
+		name: instanceMetricName(name),
+		labels: []attribute.KeyValue{
+			keyPair,
+		},
+		value: value,
+	}
+}
+
+func TestValueObservedMetrics(t *testing.T) {
+	s := scheme.Scheme
+
+	// Add jaeger to schema
+	s.AddKnownTypes(v1alpha1.GroupVersion, &v1alpha1.TempoStack{}, &v1alpha1.TempoStackList{})
+
+	// Create jaeger instances
+	tempoGCS := newTempoStackInstance(types.NamespacedName{
+		Name:      "my-tempo-gcs",
+		Namespace: "test",
+	}, v1alpha1.ManagementStateManaged, v1alpha1.ObjectStorageSecretGCS)
+
+	tempoS3 := newTempoStackInstance(types.NamespacedName{
+		Name:      "my-jaeger-s3",
+		Namespace: "test",
+	}, v1alpha1.ManagementStateManaged, v1alpha1.ObjectStorageSecretS3)
+
+	tempoOtherS3 := newTempoStackInstance(types.NamespacedName{
+		Name:      "my-jaeger-other-s3",
+		Namespace: "test",
+	}, v1alpha1.ManagementStateManaged, v1alpha1.ObjectStorageSecretS3)
+
+	tempoUnmanaged := newTempoStackInstance(types.NamespacedName{
+		Name:      "my-jaeger-azure-unmanaged",
+		Namespace: "test",
+	}, v1alpha1.ManagementStateUnmanaged, v1alpha1.ObjectStorageSecretAzure)
+
+	tempoUnmanaged2 := newTempoStackInstance(types.NamespacedName{
+		Name:      "my-jaeger-azure-unmanaged-2",
+		Namespace: "test",
+	}, v1alpha1.ManagementStateUnmanaged, v1alpha1.ObjectStorageSecretAzure)
+
+	objs := []runtime.Object{
+		&tempoGCS,
+		&tempoS3,
+		&tempoOtherS3,
+		&tempoUnmanaged,
+		&tempoUnmanaged2,
+	}
+	expected := []expectedMetric{
+		newExpectedMetric(managedMetric, attribute.String("state", string(v1alpha1.ManagementStateManaged)), 3),
+		newExpectedMetric(managedMetric, attribute.String("state", string(v1alpha1.ManagementStateUnmanaged)), 2),
+		newExpectedMetric(storageBackendMetric, attribute.String("type", string(v1alpha1.ObjectStorageSecretGCS)), 1),
+		newExpectedMetric(storageBackendMetric, attribute.String("type", string(v1alpha1.ObjectStorageSecretS3)), 2),
+		newExpectedMetric(storageBackendMetric, attribute.String("type", string(v1alpha1.ObjectStorageSecretAzure)), 2),
+	}
```

Comment:
Wouldn't it make sense to do this assertions in a loop?

Reply:
At the end this is evaluated in a loop, but this way it makes me easier to see which metrics I'm evaluating. I just fin easier to read.